## Introduction

Modular arithmetic deals with finding the remainder of an integer division.

For example, 16 and 126 have the remainder when divided by 5. In notation,

$$16 \equiv 126 \pmod{5}$$

This is a way to simplify number problems by replacing the infinite set of integers $\mathbb{Z}$ with a simpler number system $\mathbb{Z}_n$ with only $n$ numbers.

<!-- speak of congruence modulo n?
  $$\mathbb{C}_n = \{ (a,b) \mid a-b = kn, ~ ~ a,b,k \in \mathbb{Z} \}$$
-->

The properties of addition, subtraction, multiplication and exponentiation are kept, while division must be handled with some care.

For example, one property says that, in $\!\!\!\pmod{n}$, if $a \equiv r_1$ and $a \equiv r_2$, then $a+b \equiv r_1+r_2$

In [22]:
n = 21337
a,  b  = 1431524135, 674206923
r1, r2 = a % n,      b % n

assert (a+b) % n == (r1+r2) % n

If $a,b$ have the same remainder when divided by $n$, we say that $a$ and $b$ are **congruent** mod $n$.

If $a,b$ are congruent mod $n$, then $a-b$ and $b-a$ are multiples of $n$.

In [ ]:
n = 5
a, b = 54316, 132726

assert (a-b) % n == 0 and (b-a) % n == 0

An example. It is know that a perfect square $n = m^2$ is congruent, mod $4$, with either $1$ or $0$. This is because $m$ can be expressed as $m = 4q + r$, making $n = m^2 = 16q^2 + 8qr + r^2$. The remainder $r$ can only be $0,1,2,3$ and we can manually test each one, to check that the possible remainders of $n%4$ are either zero or one.

Now, is $x = 22051946$ a perfect square? We can decompose it as $22051946 = 220519 \times 100 + 46 = 220519 \times 25 \times 4 + 46$. Modulo 4, $x$ is congruent with $46$, which is also congruent with $2$, so $x$ is not a perfect square.

In notation we'll write,

$$22051946 ≡  220519 \times 25 \times 4 + 46 ≡ 46 ≡ 2 \pmod 4$$

Notice that this is a necessary condition, not a sufficient one. A counter example is, say, $82997$, which $82997 ≡ 97 ≡ 1 \pmod 4$ and is not a perfect square.

### Properties of Addition

+ If $a+b=c$ then $a\pmod{n} + b \pmod{n} = c \pmod{n}$

+ If $a\equiv b\pmod{n}$ then $a+k \equiv b+k\pmod{n}$, for any integer $k$

+ If $a\equiv b\pmod{n}$ and $c \equiv d$ then $a+c \equiv b+d\pmod{n}$

+ If $a \equiv b\pmod{n}$ then $-a \equiv -b\pmod{n}$

Example: What is the remainder of 3455 + 1234 + 6533 divided by 5?

We don't need to sum. Just find the remainders of each number and sum those:

$$3455 + 1234 + 6533 ≡ 0 + 4 + 3 ≡ 7 = 2 \pmod{5}$$

Example: Now is 19:00. What will be the time after 1000 hours?

$$19 + 1000 ≡ 19 + (24\times 41 + 16) ≡ 35 + 24 \times 41 ≡ 35 ≡ 11 \pmod{24}$$

In [ ]:
now, after = 19, 1000

assert (now + after) % 24 == 11

Example: What is the remainder of $1! + 2! +3! + \ldots + 50!$ divided by $5!$?

Since all factorials from $5$ to $50$ are multiples of $5!=120$, their remainders will be zero, so

$$1! + \ldots + 50! ≡ 1! +2!+3!+4! ≡ 1+2+6+24 = 33 \pmod{120}$$

### Properties of Multiplication

+ If $ab = c$, then $a \pmod{n} \cdot b \pmod{n} ≡ c \pmod{n}$

+ If $a \equiv b \pmod{n}$, then $ka ≡kb \pmod{n}$, for any integer $k$

+ If $a ≡ b \pmod{n}$ and $c ≡ d \pmod{n}$, then $ac ≡ bd \pmod{n}$

+ If $ka ≡ kb \pmod{kn}$, then $a ≡ b \pmod{n}$, for $k \neq 0$

Example: What is $702 \times 619 \pmod{7}$?

Since $702 ≡ 2 \pmod{7}$ and $619 ≡ 3 \pmod{7}$,

$$702 \times 619 ≡ 2 \times 3 ≡ 6 \pmod{7}$$

In [ ]:
assert 702 * 619 % 7 == 6

What is $124 \times 634 \times 34 \times 200 \pmod{3}$?

Again, it is easier to compute the remainders of each term,

$$124 \times 634 \times 34 \times 200 ≡ 1 \times 4 \times 4 \times 2 ≡ 1 \times 1 \times 1 \times 2 ≡ 2 \pmod{3}$$

In [ ]:
assert 124 * 634 * 34 * 200 % 3 == 2

Example: What is $\underbrace{111...111}_{124} \pmod{271}$?

To solve this question is necessary to check if there's some number with just ones that is a multiple of $271$:

In [ ]:
for n_ones in range(1, 124):
  n = int(str('1')*n_ones)
  if n%271==0:
    print(f"{'1'*n_ones} ({n_ones} ones)")
    break

11111 (5 ones)


We found that $11111$ is a multiple of $271$, so $11111 ≡ 0 \pmod{271}$.

Let's multiply by $10^5$,

$$11111 \times 10^5 ≡ 0 \pmod{271}$$

and add another $11111$ to both sides (but since is a multiple of $271$, $11111$ is congruent with zero),

$$\underbrace{111...111}_{10} ≡ 11111 ≡ 0 \pmod{271}$$

Just repeat this $22$ more times, and add the last four ones,

$$\underbrace{111...111}_{120} \times 10000 \times 1111 ≡ 1111 ≡ 27 \pmod{271}$$

### Properties of Exponentiation

+ If $a ≡ b \pmod{n}$, then $a^k ≡ b^k \pmod{n}$, for any integer $k$

+ If $x, y$ are positive integers with $x ≥ y$ , then $a^x ≡ 0 \pmod{a^y}$, for any integer $a$

Example: What is $3^{16} \pmod{4}$?

Notice that $3 ≡ -1 \pmod{4}$ so,

$$3^{16} ≡ (-1)^{16} ≡ 1 \pmod{4}$$

Example: What is the last digit of $17^{17}$?

$$17^{17} ≡ 7^{17} ≡ (7^2)^8 \cdot 7 ≡ 49^8 \cdot 7 ≡ 9^8 \cdot 7 ≡ (81)^4 \cdot 7 ≡ 1^4 \cdot 7 ≡ 7 \pmod{10}$$

Python `pow` function includes a third argument for a mod value:

In [ ]:
assert pow(17, 17, 10) == 7  # 17^17 mod 10

Example: Find the last two digits of $2^{40}$.

$$(2^5)^8 ≡ 32^8 ≡ (32^2)^4 ≡ 1024^4 ≡ 24^4 ≡ (24^2)^2 ≡ 576^2 ≡ 76^2 ≡ 5776 ≡76  \pmod{100}$$

In [ ]:
assert 2**40 % 100 == 76

Example: What is $2^{123456789} \pmod{7}$?

$$(2^3)^{41152263} ≡ 8^{41152263} ≡ 1^{41152263} ≡ 1 \pmod{7}$$

In [ ]:
assert pow(2, 12345789, 7) == 1

It is also possible to find repeating patterns in powers to help solve a problem.

Example: find the last digit of $7^{358}$

First let's check the last digits of $7^0, 7^1, 7^2, 7^3, ...$

In [ ]:
[7**i%10 for i in range(12)]

[1, 7, 9, 3, 1, 7, 9, 3, 1, 7, 9, 3]

There is a cycle of period 4. Since $358 \% 4 = 2$ the last digit is the third element in the cycle, i.e., $9$.

In [ ]:
assert pow(7, 358, 10) == 9

### Properties of Division

Two numbers $a,b$ are **coprimes** if $\gcd(a,b)=1$.

+ If $ka ≡ kb \pmod{n}$ and $k,n$ are coprimes, then $a ≡ b \pmod{n}$

+ If $a ≡ b \pmod{mn}$, then $a ≡ b \pmod{n}$

## Euler's Totient Function

Given integer $n$, the numbers from $\{0, n-1\}$ that are coprime to $n$ form a group under multiplication modulo $n$. That is, the operation is associative, there is an identity (the number $1$), and all elements have inverse.

This is denoted as **multiplicative group of integers modulo n**

Example: For $n=12$ the group is $(\{1,5,7,11\}, \times)$ with the following multiplication table,

 | × (mod 12) |1  | 5 | 7 |11 |
 |----------|---|---|---|---|
 | **1**| 1 | 5 | 7 |11 |  
 | **5**| 5 | 1 |11 |7  |
 | **7**| 7 |11 |1  |5  |
 |**11**|11 |7  |5  |1  |


The order of this group (i.e., the number of its elements) is given by Euler's [totient function](https://en.wikipedia.org/wiki/Euler%27s_totient_function) $\varphi(n)$,

In [ ]:
from sympy import primefactors

def totient(n):
  """ ϕ(n) = n * (1 - 1/p1) * ... (1 - 1/pn)
      given the prime factorisation of n is p1**e1 * ... * pn**en """
  phi = n
  for factor in primefactors(n):
    phi *= 1 - 1/factor
  return int(round(phi)) # round (just in case)

assert totient(12) == 4

+ If $a,b$ are coprimes, then $\varphi(ab) = \varphi(a)\varphi(b)$

+ $\varphi(p^k) = p^{k-1}(p-1)$, for prime $p$ and $k>1$

In [ ]:
a, b = 12, 43
assert totient(a*b) == totient(a) * totient(b)

p, k = 17, 3
assert totient(p**k) == p**(k-1) * (p-1)

Example: what are the number of coprimes of $400$, up to itself?

$$\varphi(400) = \varphi(2^4 5^2) = \varphi(2^4)\varphi(5^2) = 2^3 \cdot (2-1) \cdot 5^1 \cdot (5-1) = 8 \cdot 1 \cdot 5 \cdot 4 = 160$$

In [ ]:
assert totient(400) == 160

+ $\sum_{d|n} \varphi(d) = n$

where $d|n$ means $d$ is a divisor of $n$.

In [ ]:
n = 12
assert sum(totient(d) for d in range(1,n+1) if n%d==0) == n

+ If $d|n$ then $\varphi(d) | \varphi(n)$

+ $\varphi(ab) = \varphi(a) \cdot \varphi(b) \cdot \frac{d}{\varphi(d)}$, where $d=\gcd(a,b)$

   + $\varphi(a^k) = a^{k-1} \varphi(a)$

+ $\varphi(a)\varphi(b) = \varphi(\gcd(a,b)) \cdot \varphi(\text{lcm}(a,b))$   

## Euler's Theorem

The theorem states,

+ If $a,n$ are coprimes then $a^{\varphi(n)} ≡ 1 \pmod{n}$

Fermat's little theorem is a special case for when $n$ is prime (since $\varphi(p)=p-1$),

+ $a^p ≡ a \pmod{p}$

If $a,p$ are coprimes then, by the properties of division, $a^{p-1} ≡ 1 \pmod{p}$

Example: $2046$ is divisible by $11$, since $2^{11} - 2 = 2046$.

Example: if $a,x$ are positive integers greater than $2$, what is the value of $a$ given that $a^x ≡ a-2\pmod{a-1}$?

Let $b=a-1$, so $(b+1)^x ≡ b-1 ≡ -1\pmod{b}$.

Applying the binomial expansion $(b+1)^x$, all terms are multiples of $b$ except $1^x$, so the expression simplifies to $1^x ≡ -1 \pmod{b}$.

Of course, $1^x = 1$. Summing one to both terms, $2 ≡ 0 \pmod{b}$. This is only possible for $b=2$, so $a=3$.

## Multiplicative Inverses

The modular inverse of $a$ in the ring of integers modulo $n$, is an integer $x$ such that,

$$ax ≡ 1 \pmod{n}$$

This value exists if $a,n$ are coprimes.

In Python this can be found using `pow`,

In [ ]:
a, n = 150, 7
x = pow(a, -1, n)

assert a*x % n == 1

Even if a simlar `pow` function is not available, there is an efficient algorithm to compute the multiplicative inverse using the _extended Euclidean algorithm_ (see next section).

## Bézout's Identity

The Bézout's identity states:

> Let integers $a,b$ have a gcd of $d$. There exists $x,y$ such that $ax+by=d$.

Here we assume $gdc(0,0)=0$

If $a,b$ are coprimes, then $ax+by=1$.

The [extended Euclidean algorithm](https://en.wikipedia.org/wiki/Extended_Euclidean_algorithm) is used to compute $x,y$.

In [2]:
def egcd(a, b):
  """ computes gcd(a,b) and the coefficients (x,y) of Bézout's identity
      https://en.wikibooks.org/wiki/Algorithm_Implementation/Mathematics """
  if a == 0:
    return (b, 0, 1)
  g, y, x = egcd(b%a, a)
  return g, x-(b//a)*y, y

Suppose we wish to find $x,y$ such that $3002x + 1329y = 1$,

In [ ]:
a, b = 3002, 1329
g, x, y = egcd(a, b)
print(f'x: {x}, y: {y}')
assert g == 1         # gcd must be 1
assert a*x + b*y == 1 # checking Bézout's identity

x: -622, y: 1405


Example: Find the smallest positive integer $n$ such that the equation $455x+1547y=50,000+n$ has a solution $x,y$, where both $x$ and $y$ are integers.

In [ ]:
a, b = 455, 1547
g, x, y = egcd(a, b)

print('gcd:', g) # gcd(a,b) is 91
print(f'50000 % {g}: ', 50000%g)
print(f'The first 50000+n must sum to a multiple of {g}, ie, n={g-50000%g}')

gcd: 91
50000 % 91:  41
The first 50000+n must sum to a multiple of 91, ie, n=50


As said, it is possible to compute the multiplicative inverse using the extended Euclidean algorithm:

In [3]:
def mul_inv(a, n):
  g, x, _ = egcd(a, n)
  if g == 1:
    return x % n
  raise Exception('modular inverse does not exist')

a, n = 150, 7
assert mul_inv(a, n) == pow(a, -1, n)

## Chinese Remainder Theorem

Given pairwise coprime positive integers $n_1,n_2,…,n_k$ and integers $a_1,a_2,…,a_k$, the system of congruences

$$
x ≡ a_1  \pmod{n_1} \\
x ≡ a_2  \pmod{n_2} \\
\vdots \\
x ≡ a_k  \pmod{n_k}
$$

has a solution, and it is unique modulo $N = n_1n_2\ldots n_k$.

In [4]:
from math import prod

def chinese_remainder(a_s, ns):
  product   = prod(ns)
  summation = sum(a * (p := product//n) * mul_inv(p, n)
                  for a, n in zip(a_s, ns))
  return summation % product

Example: Solve this system of congruences,
$$
x ≡ 2  \pmod{3} \\
x ≡ 3  \pmod{5} \\
x ≡ 2  \pmod{7}
$$

In [ ]:
a_s, ns = [2,3,2], [3,5,7]

x = chinese_remainder(a_s, ns)

assert all(x % n == a for a, n in zip(a_s, ns))

print(f'Solution: x ≡ {x}  (mod {prod(ns)})')

Solution: x ≡ 23  (mod 105)


Sometimes we can solve these systems even when the moduli are not co-prime.

Consider
$$
x ≡ 5  \pmod{6} \\
x ≡ 3  \pmod{8}
$$

This cannot be done directly,

In [ ]:
try:
  chinese_remainder([5,3],[6,8])
except Exception:
  print('No solution found')

No solution found


However, given $\gcd(6,8)=2$, and that the both congruences implies $x ≡ 1 \pmod 2$, this means the two congruences are not contradictory, and a solution can exist.

We can divide both moduli by their gcd and get co-primes:

$$
x ≡ 2  \pmod{3} \\
x ≡ 3  \pmod{4}
$$

The solution modulo, however, will be given by $\frac{6 \times 8}{\gcd(6,8)}$, that is, $\text{lcm}(6,8)$.

In [ ]:
from math import lcm

x = chinese_remainder([2,3], [3,4])
print(f'Solution: x ≡ {x}  (mod {lcm(6,8)})')

Solution: x ≡ 11  (mod 24)


Example: The number of students in a school is between 500 and 600. If we group them into groups of 12, 20, or 36 each, 7 students are always left over. How many students are in this school?

In [ ]:
from math import gcd

print('gcd:', gcd(12, 20, 36))
# The gcd is 4. The congruences at mod 4 are all x≡3, so they are compatible

# Now, we need to make the moduli co-prime, eg:
x = chinese_remainder([3,2,7], [4,5,9])
print(f'Solution: x ≡ {x}  (mod {lcm(12,20,36)})')

gcd: 4
Solution: x ≡ 7  (mod 180)


In [ ]:
while not(500 < x < 600):
  x += lcm(12,20,36)

print(f'There are {x} students')

There are 547 students


Example: What are the last two digits of $49^{19}$?

This needs to compute $49^{19} \pmod {100}$. $100 = 4 \times 25$ which are coprime.

So this problem, according to the Chinese remainder theorem might be translated to the next system of congruences
$$
x ≡ 49^{19}  \pmod{4} \\
x ≡ 49^{19}  \pmod{25}
$$

Using the exponentiation properties,
$$
x ≡ 1^{19} ≡ 1 \pmod{4} \\
x ≡ (-1)^{19} ≡ -1  \pmod{25}
$$

In [ ]:
x = chinese_remainder([1,-1], [4,25])
print(f'Solution: x ≡ {x}  (mod {4*25})')

Solution: x ≡ 49  (mod 100)


So the last two digits of $49^{19}$ are $49$.

In [1]:
assert pow(149, 19, 100) == 49

## Wilson's Theorem

> A positive integer $n$ is a prime iff $(n-1)! ≡ -1 \pmod{n}$

Example: evaluate $30! \pmod{899}$.

Factorizing $899 = 29 \times 31$.

$30! = 30 \times 29 \times 28! ≡ 0 \pmod{29}$ (because the $29$ makes $30!$ a multiple of $29$)

$30! ≡ -1 \pmod{31}$ by Wilson's theorem.

Now we have a system of congruences:

In [5]:
x = chinese_remainder([0,-1], [29,31])
print(f'Solution: x ≡ {x}  (mod {29*31})')

Solution: x ≡ 464  (mod 899)


In [6]:
from math import factorial

assert factorial(30) % 899 == x

By dividing both sides by $n-1$ we get a corollary of Wilson's theorem,

+ if $p$ is prime, then $(p-2)! ≡ 1 \pmod p$

Example: evaluate $97! \pmod{101}$

Since $101$ is prime, $100! ≡ -1 \pmod{101}$ by Wilson's theorem, and $97! = \frac{100!}{100 \times 99 \times 98}$,

$$97! ≡ \frac{-1}{-1 \times -2 \times -3} ≡ \frac{1}{6} \pmod{101}$$

Let's find the multiplicative inverse of 6,

In [8]:
mul_inv(6, 101)

17

So, $97! ≡ \frac{1}{6} ≡ 17 \pmod{101}$

In [9]:
assert factorial(97) % 101 == 17

## References

+ [Number Theory](https://brilliant.org/courses/basic-number-theory/) course @ Brilliant.org

+ Gareth Jones, Mary Jones - Elementary Number Theory (1998)